airflow/www/views.py

```diff
@@ -231,6 +231,14 @@ def get_date_time_num_runs_dag_runs_form_data(www_request, session, dag):
     }
 
 
+def _safe_parse_datetime(v: str):
+    """Parse datetime and return error message for invalid dates"""
+    try:
+        return timezone.parse(v)
+    except ParserError:
+        abort(400, f"Invalid datetime: {v!r}")
```

Comment:
```suggestion
def _safe_parse_datetime(v: Optional[str]) -> DateTime:
    """Parse datetime and return error message for invalid dates"""
    if v is None:
        abort(400, f"datetime not provided")
    try:
        return timezone.parse(v)
    except ParserError:
        abort(400, f"Invalid datetime: {v!r}")
```

Reply:
This produces mypy error since `abort` is treated as a no return statement. I tried annotating with `Union[DateTime, NoReturn]` but it also doesn't work and some discussion on this https://github.com/python/typing/discussions/994. 

```
Run mypy for core........................................................................Failed
- hook id: mypy
- exit code: 1

airflow/www/views.py:234: error: Missing return statement  [return]
    def _safe_parse_datetime(v: Optional[str]) -> DateTime:
    ^
airflow/www/views.py:237: error: Argument 1 to "parse" has incompatible type
"Optional[str]"; expected "str"  [arg-type]
            return timezone.parse(v)
                                  ^
Found 2 errors in 1 file (checked 1 source file)

ERROR: The previous step completed with error. Please take a look at output above 
```

Reply:
Since it was a small and non-public function I left it without typing after several tries of Optional, Union and NoReturn. Suggestions welcome.

Reply:
Yes you can’t do `NoReturn` in a return annotation. But that’s likely not the cause to the error you’re getting; this one is probably because you didn’t add the None check:

```python
    if v is None:
        abort(400, f"datetime not provided")
```

But yeah it’s OK to not add the return annotation (in which case you don’t need to add the None check either) if you feel it’s too much trouble.

Reply:
Even after using `Optional[str]` and doing a check for None mypy complains that `Optional[str]` cannot be passed to `timezone.parse` which needs a str. Probably it needs a casting : https://github.com/python/typing/issues/645

The case where a function could return or call a function that raises exception is raised here : https://github.com/python/mypy/issues/8964. This also seems to require cast and seems to make the code complicated for typing alone. I have removed hints and added `TypeError` too for the function to handle non-string objects. Sometimes mypy is not so intuitive for these smaller cases :(

Similar https://github.com/python/mypy/issues/9392